### PR TITLE
Add "Account System"

### LINKED ISSUE #3: Signing in with Scratch
**Is your feature request related to a problem? Please describe.**
When new tutorials are created, they should go with the valid user who made it.

**Describe the solution you'd like**
We need a system to make sure that when creating an account, this is the actual person on scratch. We need a solution like https://ocular.jeffalo.net/login which will safely connect the scratch account to the user.

**Describe alternatives you've considered**
Not on the top of my mind, but if anyone has any, it will be considered.

### PR DESCRIPTION
**Pull request by FunctionalMetatable**
**Resolves**

Resolves #3

**Changes**

- [x] Adds an account system. **NOTE that this is just the backend**
- [x] Some changes to the README
- [x] Adds CORS (you need to set the URL to your one in `index.js`)

**Reason for changes**
Because we need it

**Tests**

via repl

### CODE REVIEW

**Comment by TheColaber** [comment]
Wow! I am impressed by the quality of your PR! However, could you split the spelling mistakes, etc. to a different PR so that way it's easier to review what is actually related to creating accounts?

**Review by TheColaber — CHANGES_REQUESTED** [review]
Looks amazing other than these things

File: index.js [hunk 1]
```diff
@@ -76,4 +76,44 @@ app.post("/api/new", (req, res) => {
...
+    return res.json({error: 'invalid token'})
+  })
+  .then(resp => {
+    if (!resp) return;
+    resp.sessions = null // remove this so we mask it away from da client
+    res.json(resp)
+  })
+})
+
+app.get("/auth/delete", (req, res) => {
```
[the remark this answers is not in the document]
Comment (FunctionalMetatable) [reply index.js:110]: just for the future

File: index.js [hunk 2]
```diff
@@ -76,4 +76,44 @@ app.post("/api/new", (req, res) => {
...
 });
 
+app.get("/login", (req, res) => {
+  res.redirect(`https://FluffyScratch.hampton.pw/auth/getKeys/v2?redirect=${Buffer.from(REPLIT_URL + "/login/finish", 'utf-8').toString("base64")}`)
+})
+
+// /login/finish - Finishes the logging-in process from FluffyScratch. It issues a cookie to the user 
+app.get("/login/finish", async (req, res) => {
+  if (!req.query.privateCode) return res.redirect(`/login/error?error=0`)
+  var username = await auth.checkIfFluffyResponseValid(req.query.privateCode).catch(err => {res.redirect(`/login/error?error=1`); throw ""})
```
[the remark this answers is not in the document]
Comment (FunctionalMetatable) [reply index.js]: now it does

File: index.js [hunk 3]
```diff
@@ -1,9 +1,12 @@
+const REPLIT_URL = "FlagClicked.9gr.repl.co"; // Please change this URL to your replit URL before starting!
```
Comment (TheColaber) [inline index.js:1]: I kinda wanted to make the url figure itself out... Maybe express has something that could tell us what domain it's running on?

File: pages/login/error.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <title>Login Error - When Flag Clicked</title>
+    <meta property="og:title" content="When Flag Clicked" />
+    <meta property="og:description" content="When Flag Clicked Home Page." />
```
Comment (TheColaber) [inline pages/login/error.html:6]: Needs to be changed later

File: index.js (hunk 3 above)
Comment (FunctionalMetatable) [reply index.js:1]: doesnt seem there is.. there is a hacky method for this

**Review by TheColaber — APPROVED** [review]

**Comment by FunctionalMetatable** [comment]
now looks done

**Review by TheColaber — CHANGES_REQUESTED** [review]

File: pages/index.html [hunk 5]
```diff
@@ -14,6 +14,17 @@
...
+          window.location.searchParams = new URLSearchParams(
+            window.location.search
+          );
+          if (location.searchParams.get("clearCookie")) {
+            fetch("/login/delete", { method: "GET", credentials: "include" }); // invalidate the token to prevent spam. just in case
+            var cookies = document.cookie.split("; ");
+            cookies.forEach((cookie) => {
+              document.cookie += ";max-age=0;";
+            });
+            window.location.replace("/");
```
[the remark this answers is not in the document]
Comment (FunctionalMetatable) [reply pages/index.html]: just trying to fix the cookie thing before... its fixed now

File: index.js [hunk 6]
```diff
@@ -60,20 +63,85 @@ app.get("/tutorials/:id", (req, res) => {
...
+
+  auth
+    .getSession(cookie)
+    .catch((err) => {
+      res.status(403).json({ error: "invalid token" });
+      throw "";
+    })
+    .then((res) => {
+      db.list("tutorial-").then((matches) => {
+        db.set(`tutorial-${matches.length + 1}`, req.body).then(() => {
```
[the remark this answers is not in the document]
Comment (FunctionalMetatable) [reply index.js]: done

**Review by TheColaber — APPROVED** [review]

**Comment by FunctionalMetatable** [comment]
replit broken it says someone named ttury has committed to the repository but they haven't lol

**Comment by FunctionalMetatable** [comment]
@TheColaber you might want to wipe the tutorials on the replit db and stuff so the new tutorials can work properly

**Pull request merged**